### PR TITLE
usb: changes to permit endpoint addr reuse across alternates

### PR DESCRIPTION
**Pull request by mparanjpe**
Two changes to allow endpoint address reuse across alternate settings of the same interface:

* In `subsys/usb/device/usb_descriptor.c`:

    The function `usb_fix_descriptor()` keeps track of endpoints which have already been used and modifies entries in interface descriptors which use the same endpoints. However, this logic prevents reuse of endpoints across alternate settings of the same interface.

    Modify the `usb_fix_descriptor()` logic to allow reuse of endpoints across alternate settings of the same interface.

* In `subsys/usb/device/usb_device.c`:

    The logic in `usb_set_interface()` iterates over endpoint definitions in all the interfaces and appropriately configures the endpoints it finds. If a particular endpoint address appears twice in the same interface, but in different alternates, then that endpoint address must only be configured when the chosen alternate setting is being acted on.

    For example, if OUT endpoint address 1 appears in alternate settings 0 and 1, then it will appear in the endpoint descriptors corresponding to both the interface's alternate settings. Next, suppose that the chosen alternate setting is 1. For this situation, `usb_set_interface()` must only call `usb_eps_reconfigure()` when iterating over endpoint 1 as part of the chosen alternate setting 1. When `usb_set_interface()` encounters endpoint 1 while iterating over alternate setting 0, `usb_eps_reconfigure()` should not be called on endpoint 1.

    This commit accomplishes the above by:

    1. First accumulating a bit-mask of all the endpoints used by the selected interface and alternate-setting combination

    2. Before calling `usb_eps_reconfigure()` on an endpoint which isn't in the currently chosen alternate setting's descriptor, ensuring that this endpoint isn't part of the chosen alternate setting, and that the endpoint hasn't been reconfigured before. (Note that `usb_eps_reconfigure()` is called on all endpoints in the selected alternate setting)

### CODE REVIEW

**Review by jfischer-no — CHANGES_REQUESTED** [review]
tl;dr The issues are known and since there are many of them, it was decided to rework it completely. Device part of the new USB support should come into the tree in the next weeks. I would not like to change anything in this "legacy" code that is not bug-fix.

**Comment by mparanjpe** [comment]
@jfischer-no 
> tl;dr The issues are known and since there are many of them, it was decided to rework it completely. Device part of the new USB support should come into the tree in the next weeks. I would not like to change anything in this "legacy" code that is not bug-fix.

Sounds good. A big part of my motivation was to share what I had to do get my application to work. I can close this PR if you feel the underlying issue will be addressed with the reworked code. I'll also look at the branch.

File: subsys/usb/device/usb_device.c [hunk 1]
```diff
@@ -703,8 +812,13 @@ static bool usb_set_interface(struct usb_setup_packet *setup)
 	uint8_t cur_alt_setting = 0xFF;
 	uint8_t cur_iface = 0xFF;
 	bool ret = false;
+	uint32_t ep_mask_by_alt[MAX_ALTS_PER_IFACE];
```
[the remark this answers is not in the document]
Comment (mparanjpe) [reply subsys/usb/device/usb_device.c:815]: This is an excellent point - I don't know why I didn't write the code to only make a note of the endpoints in the selected or chosen alt-setting. Let me rewrite to eliminate the array

File: subsys/usb/device/usb_device.c [hunk 2]
```diff
@@ -688,6 +724,79 @@ static bool usb_set_configuration(struct usb_setup_packet *setup)
...
+ * endpoints appropriately based on alternate setting. At the moment
+ * this means that for an endpoint that is used in the currently
+ * chosen alternate setting, set_endpoint() is called; while if the
+ * currently chosen alternate setting does not use an endpoint used
+ * elsewhere in the same interface's descriptor, reset_endpoint() is
+ * called on this unused endpoint.
+ *
+ * For the situation where multiple alternates of the selected
+ * interface use the same endpoint address, the logic needs to have an
+ * enumeration of all the endpoints used by all the alternate settings
```
[the remark this answers is not in the document]
Comment (mparanjpe) [reply subsys/usb/device/usb_device.c:740]: What I'm trying to say is that before iterating through the USB interface descriptors to call `usb_eps_reconfigure()`, the logic needs to a-priori know all the endpoints used by the selected or chosen alt-setting.

File: subsys/usb/device/usb_device.c [hunk 3]
```diff
@@ -726,8 +840,22 @@ static bool usb_set_interface(struct usb_setup_packet *setup)
 		case USB_DESC_ENDPOINT:
 			if (cur_iface == setup->wIndex) {
 				ep = (struct usb_ep_descriptor *)p;
-				ret = usb_eps_reconfigure(ep, cur_alt_setting,
-							  setup->wValue);
+
+				/*
+				 * if endpoint will be configured by alt-setting,
```
[the remark this answers is not in the document]
Comment (mparanjpe) [reply subsys/usb/device/usb_device.c:845]: Maybe I should reword this comment. I'm trying to say that `usb_eps_reconfigure()` should only be called on an endpoint when `cur_alt_setting` is the same as the selected or chosen `alt_setting`

File: subsys/usb/device/usb_device.c [hunk 4]
```diff
@@ -688,6 +724,79 @@ static bool usb_set_configuration(struct usb_setup_packet *setup)
...
+static void get_ep_masks_for_alts(uint8_t iface, uint32_t *ep_mask,
+				  uint8_t num_alt)
+{
+	const uint8_t *p = usb_dev.descriptors;
+	struct usb_ep_descriptor *ep = NULL;
+	uint8_t cur_alt_setting = 0xff;
+	uint8_t cur_iface = 0xff;
+
+	/* clear endpoint bitmask array by alt-setting */
+	memset(ep_mask, 0, num_alt * sizeof(ep_mask[0]));
```
[the remark this answers is not in the document]
Comment (mparanjpe) [reply subsys/usb/device/usb_device.c:764]: Yes - I'll be redoing the functionality to remove the array...

File: subsys/usb/device/usb_descriptor.c [hunk 5]
```diff
@@ -374,7 +374,9 @@ static int usb_fix_descriptor(struct usb_desc_header *head)
 	struct usb_ep_descriptor *ep_descr = NULL;
 	uint8_t numof_ifaces = 0U;
 	uint8_t str_descr_idx = 0U;
-	uint32_t requested_ep = BIT(16) | BIT(0);
+	uint32_t requested_ep_iface = BIT(16) | BIT(0);
+	uint32_t requested_ep_accum = requested_ep_iface;
+	uint32_t requested_ep_iface_alt = requested_ep_accum;
```
[the remark this answers is not in the document]
Comment (mparanjpe) [reply subsys/usb/device/usb_descriptor.c]: I have renamed the variables

File: subsys/usb/device/usb_device.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (mparanjpe) [reply subsys/usb/device/usb_device.c]: I have reworded the comment and inverted the logic - please take a look

File: subsys/usb/device/usb_device.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (mparanjpe) [reply subsys/usb/device/usb_device.c]: I have redone the change to eliminate the array - please take a look

File: subsys/usb/device/usb_device.c [hunk 6]
```diff
@@ -688,6 +719,56 @@ static bool usb_set_configuration(struct usb_setup_packet *setup)
...
+ *        [in] alt_setting  The alternate setting of above interface for
+ *                          which endpoint bit-mask should be accumulated
+ *
+ * @return The accumulated 32-bit endpoint bit-mask
+ */
+static uint32_t get_ep_bm_for_alt_setting(uint8_t iface, uint8_t alt_setting)
+{
+	const uint8_t *p = usb_dev.descriptors;
+	struct usb_ep_descriptor *ep = NULL;
+	uint8_t cur_alt_setting = 0xff;
```
Comment (desowin) [inline subsys/usb/device/usb_device.c:741]: The reason behind the magic numbers is to skip endpoint descriptors that are not within requested interface alternate setting. The values `0xff` are extremely unlikely, but are not out of range. I think the purpose behind the variables could be clearer if both `cur_alt_setting` and `cur_iface` were replaced with `bool alt_setting_match = false;`. Then `alt_setting_match` would be set to `true` if descriptor bInterfaceNumber/bAlternateSetting matches requested values (and to `false` on mismatch) instead of assigning the `cur_alt_setting` and `cur_iface`.

File: subsys/usb/device/usb_device.c [hunk 7]
```diff
@@ -703,9 +784,13 @@ static bool usb_set_interface(struct usb_setup_packet *setup)
 	uint8_t cur_alt_setting = 0xFF;
 	uint8_t cur_iface = 0xFF;
 	bool ret = false;
+	uint32_t ep_bm_active;
+	uint32_t ep_bm;
 
 	LOG_DBG("Set Interface %u alternate %u", setup->wIndex, setup->wValue);
 
+	ep_bm_active = get_ep_bm_for_alt_setting(setup->wIndex, setup->wValue);
```
Comment (desowin) [inline subsys/usb/device/usb_device.c:792]: I believe this line refers to `1. First enumerating all the endpoints used across all alternate settings of the chosen interface` in the commit description. However, there is discrepancy between the description and what the code does. I believe the description should be changed to `1. First enumerating all the endpoints used in selected alternate settings of the chosen interface`.

File: subsys/usb/device/usb_device.c [hunk 8]
```diff
@@ -726,8 +811,18 @@ static bool usb_set_interface(struct usb_setup_packet *setup)
...
+
+				/*
+				 * only reconfigure endpoint if either:
+				 * a) currently iterating over selected alt_setting
+				 * b) endpoint not used in selected alt_setting
+				 */
+				ep_bm = get_ep_bm_from_addr(ep->bEndpointAddress);
+				if (cur_alt_setting == setup->wValue ||
+				    (ep_bm_active & ep_bm) == 0) {
+					ret = usb_eps_reconfigure(
```
Comment (desowin) [inline subsys/usb/device/usb_device.c:823]: Should there be also a bitmap of endpoints already reconfigured? If there are three alternate settings, the `usb_eps_reconfigure()` can be called twice on endpoint not used in selected alternate setting.

**Review by desowin — CHANGES_REQUESTED** [review]

File: subsys/usb/device/usb_device.c [hunk 9]
```diff
@@ -78,6 +78,8 @@ LOG_MODULE_REGISTER(usb_device);
 
 #define MAX_DESC_HANDLERS           4 /** Device, interface, endpoint, other */
 
+#define MAX_ALTS_PER_IFACE          8 /** Max alternate settings per interface */
```
[the remark this answers is not in the document]
Comment (desowin) [reply subsys/usb/device/usb_device.c]: Discard this. I had this comment pending (didn't finish review when I first looked at it) and the comment "appears as new" when I submitted my review now.

File: subsys/usb/device/usb_descriptor.c [hunk 10]
```diff
@@ -407,6 +410,8 @@ static int usb_fix_descriptor(struct usb_desc_header *head)
 				}
 			}
 
+			ep_bm_active = ep_bm_iface;
+			ep_bm_tmp = ep_bm_active;
```
[the remark this answers is not in the document]
Comment (mparanjpe) [reply subsys/usb/device/usb_descriptor.c]: Done

File: subsys/usb/device/usb_device.c (hunk 6 above)
Comment (mparanjpe) [reply subsys/usb/device/usb_device.c]: Thanks - changed

File: subsys/usb/device/usb_device.c (hunk 7 above)
Comment (mparanjpe) [reply subsys/usb/device/usb_device.c:813]: Are you perhaps referring to the older function description? The current description for `get_ep_bm_for_alt_setting()` reads as follows:

>Iterate through the USB interface descriptor and accumulate
>a bit-mask representing all the endpoints used by by the specified
>interface and alternate-setting combination. OUT endpoints are
>represented by the lower 16 bit-positions, while the upper 16
>bit-positions indicate IN endpoints

File: subsys/usb/device/usb_device.c (hunk 8 above)
Comment (mparanjpe) [reply subsys/usb/device/usb_device.c:846]: Good point - thanks - added

File: subsys/usb/device/usb_device.c (hunk 7 above)
Comment (desowin) [reply subsys/usb/device/usb_device.c:813]: No, I am referring to the commit description itself (not to the code): https://github.com/zephyrproject-rtos/zephyr/pull/51824/commits/7f59e76babdcc1a5f7d4d7d12aae9a843d4ebed6
Unfortunately the link will no longer function after you update it due to GitHub review system limitations.

File: subsys/usb/device/usb_device.c [hunk 11]
```diff
@@ -776,8 +832,21 @@ static bool usb_set_interface(struct usb_setup_packet *setup)
 		case USB_DESC_ENDPOINT:
 			if (cur_iface == setup->wIndex) {
 				ep = (struct usb_ep_descriptor *)p;
-				ret = usb_eps_reconfigure(ep, cur_alt_setting,
-							  setup->wValue);
+
+				/*
+				 * only reconfigure endpoint if:
+				 * a) it hasn't previously been reconfigured; AND
```
Comment (desowin) [inline subsys/usb/device/usb_device.c:838]: nitpick: a) is not indented the same as b) and c)

**Review by desowin — COMMENTED** [review]
Please fix the commit description, otherwise it is good to go.

File: subsys/usb/device/usb_device.c (hunk 7 above)
Comment (mparanjpe) [reply subsys/usb/device/usb_device.c:813]: Oops - apologies - I changed the commit message and the PR description - please take a look

File: subsys/usb/device/usb_device.c (hunk 11 above)
Comment (mparanjpe) [reply subsys/usb/device/usb_device.c:838]: Changed (the extra indent was actually intentional to show priority; but either way is OK...)

Comment (desowin) [reply subsys/usb/device/usb_device.c:838]: Oh, I didn't notice that. Now, that you mentioned the priority, I see it. It essentially is `a) AND [b) or c)]`.
So maybe
```
    * only reconfigure endpoint if it hasn't previously been reconfigured AND
    * a) currently iterating over selected alt_setting; OR
    * b) endpoint not used in selected alt_setting
```
can be clearer. However, it is up to you.

**Review by desowin — DISMISSED** [review]

File: subsys/usb/device/usb_device.c (hunk 11 above)
Comment (mparanjpe) [reply subsys/usb/device/usb_device.c]: I like it - changed...

**Review by desowin — APPROVED** [review]
